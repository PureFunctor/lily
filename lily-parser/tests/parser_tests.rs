use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
