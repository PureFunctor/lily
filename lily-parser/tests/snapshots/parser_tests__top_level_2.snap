---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 38,
                end: 67,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 48,
                        end: 67,
                        kind: BinaryOperator(
                            Expression {
                                begin: 48,
                                end: 63,
                                kind: BinaryOperator(
                                    Expression {
                                        begin: 48,
                                        end: 49,
                                        kind: Variable(
                                            "a",
                                        ),
                                    },
                                    "+",
                                    Expression {
                                        begin: 52,
                                        end: 63,
                                        kind: BinaryOperator(
                                            Expression {
                                                begin: 52,
                                                end: 53,
                                                kind: Variable(
                                                    "b",
                                                ),
                                            },
                                            "*",
                                            Expression {
                                                begin: 56,
                                                end: 63,
                                                kind: Parenthesized(
                                                    Expression {
                                                        begin: 57,
                                                        end: 62,
                                                        kind: BinaryOperator(
                                                            Expression {
                                                                begin: 57,
                                                                end: 58,
                                                                kind: Variable(
                                                                    "c",
                                                                ),
                                                            },
                                                            "+",
                                                            Expression {
                                                                begin: 61,
                                                                end: 62,
                                                                kind: Variable(
                                                                    "d",
                                                                ),
                                                            },
                                                        ),
                                                    },
                                                ),
                                            },
                                        ),
                                    },
                                ),
                            },
                            "+",
                            Expression {
                                begin: 66,
                                end: 67,
                                kind: Variable(
                                    "e",
                                ),
                            },
                        ),
                    },
                ),
            },
        ],
    },
)
